src/machi_file_proxy.erl

```diff
@@ -0,0 +1,778 @@
+%% conceit at the heart of Machi's design.
+%%
+%% Read, write and append requests for a single file will be managed 
+%% through this proxy.  Clients can also request syncs for specific
+%% types of filehandles.
+%%
+%% As operations are requested, the proxy keeps track of how many
+%% operations it has performed (and how many errors were generated.)
+%% After a sufficient number of inactivity, the server terminates 
+%% itself.
+%%
+%% TODO:
+%% 1. Some way to transition the proxy into a wedged state that 
+%% doesn't rely on message delivery.
+%%
+%% 2. Check max file size on appends. Writes we take on faith we can 
+%% and should handle.
+%%
+%% 3. Async checksum reads on startup.
+
+-module(machi_file_proxy).
+-behaviour(gen_server).
+
+-include("machi.hrl").
+
+%% public API
+-export([
+    start_link/2,
+    stop/1,
+    sync/1,
+    sync/2,
+    read/3,
+    write/3,
+    write/4,
+    append/2,
+    append/4
+]).
+
+%% gen_server callbacks
+-export([
+    init/1,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2,
+    terminate/2,
+    code_change/3
+]).
+
+-define(TICK, 30*1000). %% XXX FIXME Should be something like 5 seconds
+-define(TICK_THRESHOLD, 5). %% After this + 1 more quiescent ticks, shutdown
+-define(TIMEOUT, 10*1000).
+-define(TOO_MANY_ERRORS_RATIO, 50).
+
+-type op_stats()      :: { Total  :: non_neg_integer(), 
+                           Errors :: non_neg_integer() }.
+
+-type byte_sequence() :: { Offset :: non_neg_integer(),
+                           Size   :: pos_integer()|infinity }.
+
+-record(state, {
+    data_dir              :: string() | undefined,
+    filename              :: string() | undefined,
+    data_path             :: string() | undefined,
+    wedged = false        :: boolean(),
+    csum_file             :: string()|undefined,
+    csum_path             :: string()|undefined,
+    eof_position = 0      :: non_neg_integer(),
+    unwritten_bytes = []  :: [byte_sequence()],
+    data_filehandle       :: file:filehandle(),
+    csum_filehandle       :: file:filehandle(),
+    tref                  :: reference(), %% timer ref
+    ticks = 0             :: non_neg_integer(), %% ticks elapsed with no new operations
+    ops = 0               :: non_neg_integer(), %% sum of all ops
+    reads = {0, 0}        :: op_stats(),
+    writes = {0, 0}       :: op_stats(),
+    appends = {0, 0}      :: op_stats()
+}).
+
+%% Public API
+
+% @doc Start a new instance of the file proxy service. Takes the filename 
+% and data directory as arguments. This function is typically called by the
+% `machi_file_proxy_sup:start_proxy/2' function.
+-spec start_link(Filename :: string(), DataDir :: string()) -> any().
+start_link(Filename, DataDir) ->
+    gen_server:start_link(?MODULE, {Filename, DataDir}, []).
+
+% @doc Request to stop an instance of the file proxy service.
+-spec stop(Pid :: pid()) -> ok.
+stop(Pid) when is_pid(Pid) ->
+    gen_server:call(Pid, {stop}, ?TIMEOUT).
```

Comment:
I'm thinking that this timeout regime isn't flexible enough ... but let's skip it until there is a demonstrable problem.